source/validate_capability.cpp

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Ensures type declarations are unique unless allowed by the specification.
+
+#include "validate.h"
+
+#include <cassert>
+#include <unordered_set>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+bool IsSupportGuaranteedVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityMatrix:
+    case SpvCapabilityShader:
+    case SpvCapabilityInputAttachment:
+    case SpvCapabilitySampled1D:
+    case SpvCapabilityImage1D:
+    case SpvCapabilitySampledBuffer:
+    case SpvCapabilityImageBuffer:
+    case SpvCapabilityImageQuery:
+    case SpvCapabilityDerivativeControl:
+      return true;
+  }
+  return false;
+}
+
+bool IsSupportOptionalVulkan_1_0(uint32_t capability) {
+  switch (capability) {
+    case SpvCapabilityGeometry:
+    case SpvCapabilityTessellation:
+    case SpvCapabilityFloat64:
+    case SpvCapabilityInt64:
+    case SpvCapabilityInt16:
+    case SpvCapabilityTessellationPointSize:
+    case SpvCapabilityGeometryPointSize:
+    case SpvCapabilityImageGatherExtended:
+    case SpvCapabilityStorageImageMultisample:
+    case SpvCapabilityUniformBufferArrayDynamicIndexing:
+    case SpvCapabilitySampledImageArrayDynamicIndexing:
+    case SpvCapabilityStorageBufferArrayDynamicIndexing:
+    case SpvCapabilityStorageImageArrayDynamicIndexing:
+    case SpvCapabilityClipDistance:
+    case SpvCapabilityCullDistance:
+    case SpvCapabilityImageCubeArray:
+    case SpvCapabilitySampleRateShading:
+    case SpvCapabilitySparseResidency:
+    case SpvCapabilityMinLod:
+    case SpvCapabilitySampledCubeArray:
+    case SpvCapabilityImageMSArray:
+    case SpvCapabilityStorageImageExtendedFormats:
+    case SpvCapabilityInterpolationFunction:
+    case SpvCapabilityStorageImageReadWithoutFormat:
+    case SpvCapabilityStorageImageWriteWithoutFormat:
+    case SpvCapabilityMultiViewport:
+      return true;
+  }
+  return false;
+}
+
+}  // namespace
+
+// Validates that capability declarations use operands allowed in the current
+// context.
+spv_result_t CapabilityPass(ValidationState_t& _,
+                           const spv_parsed_instruction_t* inst) {
+  const SpvOp opcode = static_cast<SpvOp>(inst->opcode);
+  if (opcode != SpvOpCapability)
+    return SPV_SUCCESS;
+
+  if (inst->num_operands != 1) {
```

Comment:
I believe you don't need this check.
if you have something like `OpCapability Shader Foo`, the binary parser will complain with a message like `Expected next instruction, but found Foo`.  I think the binary parser will also complain if no argument is provided.
Give it a try.
If it is possible to trigger the check, you should add a unit test for it. If not, you can remove the check :)